core/src/test/scala/unit/kafka/log/LogManagerTest.scala

```diff
@@ -95,6 +96,33 @@ class LogManagerTest {
     log.appendAsLeader(TestUtils.singletonRecords("test".getBytes()), leaderEpoch = 0)
   }
 
+  @Test
+  def testCreateLogWithLogDirFallback() {
+
+    // Configure a number of directories that will fail, and one that won't
+    // making it unlikely that the good log directory will be first choice.
+    val dirs = (0 to 100)
+      .map(_.toString)
+      .map(logDir.toPath.resolve(_).toFile)
+    val goodDir = dirs(88)
+
+    logManager.shutdown()
+    logManager = createLogManager(dirs)
+    logManager.startup()
+
+    // To simulate disk failure, delete log directories and replace them with files
+    // so that Files.createDirectories will fail.
```

Comment:
I guess another way to do this might be extend `Log` and override `createLogDirectory`. Also, to remove the need for 100 directories, perhaps we could make the sorting more deterministic. For example, maybe we can sort by log count first and name second?

Reply:
That's a good idea. Even with the 100 directories, with some weird sorting it could still fail.
I'd like to avoid adding further sorting work just for the sake of testing. I think I've got a way to do it reliably without extra sorting. Looking forward to your thoughts on it.